packages/common/async/src/backpressure-executor.ts

```diff
@@ -0,0 +1,173 @@
+//
+// Copyright 2025 DXOS.org
+//
+
+import { type Context, Resource } from '@dxos/context';
+import { invariant } from '@dxos/invariant';
+
+import { DeferredTask } from './task-scheduling';
+import { sleepWithContext, asyncTimeout } from './timeout';
+
+const DEFAULT_TIMEOUT = 30_000;
+
+export type BackpressureExecutorParams = {
+  /**
+   * The time budget for the task execution.
+   * in [ms].
+   */
+  budget: number;
+
+  /**
+   * The time period for the budget calculation.
+   * in [ms].
+   */
+  budgetPeriod: number;
+
+  /**
+   * The rest time if the time budget is exceeded.
+   * in [ms].
+   */
+  restTime: number;
+
+  /**
+   * The maximum number of tasks to run in parallel.
+   */
+  maxParallelTasks: number;
+
+  /**
+   * Clean executed tasks after time period.
+   * in [ms].
+   */
+  cleanUpAfter: number;
+};
+
+export type Task<T> = {
+  run: () => Promise<T>;
+  result: Promise<T>;
+  resolve: (value: T) => void;
+  reject: (reason?: any) => void;
+
+  /**
+   * The start timestamp of the task.
+   */
+  start: number;
+
+  /**
+   * The end timestamp of the task.
+   */
+  end: number;
+
+  /**
+   * The elapsed time of the task.
+   * in [ms].
+   */
+  elapsed: number;
+
+  /**
+   * The timeout for the task.
+   * in [ms].
+   */
+  timeout: number;
+};
+
+export type RunParams<T> = {
+  run: Task<T>['run'];
+  timeout?: Task<T>['timeout'];
+}[];
+
+export class BackpressureExecutor<T> extends Resource {
+  private readonly _executionQueue: Array<Task<T>> = [];
+  private _processedTasks: Array<Task<T>> = [];
+  private _executor?: DeferredTask = undefined;
+
+  constructor(private readonly _params: BackpressureExecutorParams) {
+    super();
+  }
+
+  protected override async _open(ctx: Context): Promise<void> {
+    this._executor = new DeferredTask(ctx, async () => {
+      // Clean up old tasks.
+      this._processedTasks = this._processedTasks.filter((task) => task.end > Date.now() - this._params.cleanUpAfter);
```

Comment:
calls Date.now() repeatedly. Getting the time does have a CPU impact, so please pull it out from the loop